src/biof.1
.TH "RSS Other Screensaver" 1 "" "X Version 11"
.de Ds
.Sp
.nf
..
.de De
.fi
..
.SH NAME
biof - rotating stack of quads.
.SH SYNOPSIS
.B biof
[\--root/-r]
[\--maxfps/-x \fInumber\fP]
[\--vsync/-y \fInumber\fP]
[\--dpms/-M \fInumber\fP]
[\--preset/-P \fInumber\fP]
[\--lines/-l \fInumber\fP]
[\--points/-p \fInumber\fP]
[\--offangle]
.SH DESCRIPTION
This is an attempt to recreate some of the work of William Latham. Take a look at the program 'organic art' at this site (http://www.artworks.co.uk).

Ported to Linux by Tugrul Galatali.
.SH OPTIONS
.TP 8
.B \--root
Draw on the root window.
.TP 8
.B \--maxfps \fInumber\fP
Set maximum frame rate.
.TP 8
.B \--vsync \fInumber\fP
Limit redraws to specified number of vertical refreshes.  0 - 100.  Default: 1
.TP 8
.B \--dpms \fInumber\fP
Stop rendering new frames if the display is not on.  0 - 1.  Default: 1
.TP 8
.B \--preset \fInumber\fP
.Ds
1 - Triangles (--triangles)
2 - Spheres (--spheres)
3 - Big Spheres (--bigspheres)
4 - Points (--points)
5 - Lightmap (--lightmap)
.De

Sets geometry type and resets --lines and --points to the defaults from the original program.  1 - 5.  Default: 1
.TP 8
.B \--lines \fInumber\fP
Set number of lines ("arms").  1 - 1024.
.TP 8
.B \--points \fInumber\fP
Set number of points per line.  1 - 4096.
.TP 8
.B \--offangle
Interesting 45 degree view instead of default head on view.
.SH ENVIRONMENT
.PP
.TP 8
.B DISPLAY
to get the default host and display number.
.SH SEE ALSO
.BR X (1),
.SH COPYRIGHT
Copyright \(co 2002 by <hk@dgmr.nl> and Tugrul Galatali.  

BioF is free software; you can redistribute it and/or modify
it under the terms of the GNU General Public License version 2 as
published by the Free Software Foundation.

BioF is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.
.SH AUTHOR
<hk@dgmr.nl> and Tugrul Galatali <tugrul@galatali.com>



ChangeLog
******************************* Version 0.9.1  *********************************

	Added port of Pixel City, originally by Shamus Young,
http://code.google.com/p/pixelcity/
	Replaced methods deprecated by ImageMagick.

******************************* Version 0.9.0  *********************************

	Added port of drempels, originally by Ryan M. Geiss,
http://www.geisswerks.com/about_drempels.html
	Added new saver feedback by Tugrul Galatali.
	Added support for DPMS to idle the saver when the display isn't on.
	Fixed brown paper bag bug in hyperspace.
	Added two new presets to flocks inspired by Matt Richard's Wavy Gravy on
openprocessing.org
	Tweaked hufo_smoke to avoid flat flame effect.
	Fixed unknown argument segfault in lattice.
	Switched matrixview's main character array to vertex arrays, which helps
significantly on older nVidia hardware.
	hyperspace and drempels have command line arguments to disable use of
shaders in case they behave unexpectedly.
	Minor changes to fix builds on other platforms.

******************************* Version 0.8.3 *********************************

	New screen saver Lorenz contributed by Soeren Sonnenburg <sonne@debian.org>
	Fix for SF #2673397 by changing =='s to >='s in widthChange update logic,
otherwise non-sense values for interpolation would wreak havoc. Also changed for
xyzChange.

******************************* Version 0.8.2 *********************************

	Limit frame rate to vertical refresh by default if compiled with GLEW
and GLX_SGI_swap_control is available. Otherwise, maxfps set to 30.
	--nice eliminated and made default. nanosleep/usleep are much better 
behaved these days.
	spirographx modified to use vertex arrays and GNU extension sincosf if
available, dropping CPU usage significantly.
	Minor updates to correct issues noted by newer compilers.
	--disable-debug is now the default in configure.

******************************* Version 0.8.1 *********************************

	Pulled in vroot.h patch from Ari Pollak's Debian package.
	Added missing #include <assert.h> to hyperspace.cpp
	Image loading in matrixview was hanging for whatever reason,
		rewritten with the new Magick Wand interface.
	ImageMagick now required.
	timeval to double conversion now fixed with an explicit cast of tv_sec
		to double.
	Fixed automake files to use CLEANFILES
	Dropped SSE/3dnow stuff
	Fixed hyperspace.xml typo in dist_kdessconfig_DATA

******************************* Version 0.8.0 *********************************

	Rough sync against new Windows release, rssavers 0.1
		Except for skyrocket
		Added new saver Hyperspace

******************************* Version 0.7.6 *********************************

	Fix to matrixview.c where the body of a thread was accidentally left as a
subfunction of a function that exits while the thread is still alive.

	Added an autoconf check for ExportImagePixels since it was very recently
added (5.5.7).

******************************* Version 0.7.5 *********************************

	Removed multiline strings that are no longer supported by gcc 3.3.

	Added spirographx, a port of an OSX app/screensaver SpirographX by 
Holmes Futrell <holmes@neatosoftware.com>.
	http://homepage.mac.com/solidmag/spiro/

	Added matrixview, based on Alex Zolotov's Matrix GL. Hacked in external 
image loading and ability to modify the resolution of the matrix.

******************************* Version 0.7.4 *********************************

	Fixed bug in lattice when the variable passed to acos would be slightly
greater than 1, causing cascading nan's that made the program useless.

	Explicitly clearing buffers to avoid garbage in blurring modes.

	Lots of stupid little bug fixes.

******************************* Version 0.7.3 *********************************

	Added trails and blur options to flocks. Also added presets to spruce
things up in combination with random preset selection when running 
unconfigured.

	Fixed another cpu detection bug, damn off by one errors :)

	Added support for decimal integer arguments to -window-id. 

	Added .desktop files for preliminary support of KDE. Not sure where
the default directory is, but if it isn't specified by --with-kdessconfigdir,
the configure script checks for the existence of:

/usr/share/applnk/System/ScreenSavers

******************************* Version 0.7.2 *********************************

	Fixed bug in cpu detection code.

	Fixed a bug that included audio data in skyrocket when OpenAL wasn't 
found.

******************************* Version 0.7.1 *********************************

	Fixed a typo crashing skyrocket on big endian machines. Spotted by
Michel Danzer.

	Two new savers: biof and busyspheres.

	New texture smooth for colorfire.

	Fixed a bug that left screensavers running after they should have quit.

	Replaced compile time cpu feature detection with runtime detection.

******************************* Version 0.7.0 *********************************

	Extracted all binary data from the source into binary files. The
source distribution has gotten notably smaller as a result. 

	The binary files are now converted during compilation to significantly 
larger source files with escaping instead of base64 type inefficient strings.
The resulting executables have no unnecessary bloat.

	If the bzip2 library is detected on the system, it is used to compress
the binary data before conversion.

	Added --disable-bzip2 flag to configure. This is to give up space in
order to avoid a delay while decompressing, which is noticeable in skyrocket
that has about 3MB of compressed textures and audio.

	Added --disable-sound flag to configure. This chops off about 3.3MB
(before compression) of audio from skyrocket and any runtime overhead
associated with sound. Useful if you intend to keep passing volume = 0 to
skyrocket.

	Added SSE/3DNow optimized code that gives euphoria a nice 10% boost
on my P4, probably a bit more on an Athlon but I haven't had recent console 

src/biof.c
/*
 * Copyright (C) 2002  <hk@dgmr.nl>
 * Ported to Linux by Tugrul Galatali <tugrul@galatali.com>
 *
 * BioF is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License version 2 as 
 * published by the Free Software Foundation.
 *
 * BioF is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
 */

// BioF screen saver

#include <math.h>
#include <stdio.h>
#include <stdlib.h>
#include <GL/gl.h>
#include <GL/glu.h>

#include "driver.h"
#include "rsRand.h"

const char *hack_name = "BioF";

#define TRIANGLES 1
#define SPHERES 2
#define BIGSPHERES 3
#define POINTS 4
#define LIGHTMAP 5

int NRPOINTS, NRLINES;
int dGeometry;
int dOffAngle;

GLUquadricObj *shapes;

void Horn (int ribs, double bend, double stack, double twist, double twisttrans, double grow)
{
	int i;

	for (i = 0; i < ribs; i++) {
		glPushMatrix ();
		glRotated (i * bend / ribs, 1, 0, 0);
		glTranslated (0, 0, i * stack / ribs);
		glTranslated (twisttrans, 0, 0);
		glRotated (i * twist / ribs, 0, 0, 1);
		glTranslated (-twisttrans, 0, 0);

		switch (dGeometry) {
		case TRIANGLES:
			glBegin (GL_TRIANGLES);
			glColor4d (0.9, 0.8, 0.5, 1);
			glVertex3d (2, 2, 0);
			glColor4d (0.9, 0.8, 0.5, 0.0);
			glVertex3d (0, 2, 0);
			glVertex3d (2, 0, 0);
			glEnd ();

			break;

		case BIGSPHERES:
			gluQuadricDrawStyle (shapes, GLU_FILL);
			gluQuadricNormals (shapes, GLU_SMOOTH);
			gluSphere (shapes, 7 * exp (i / ribs * log (grow)), 20, 10);

			break;

		case SPHERES:
			gluQuadricDrawStyle (shapes, GLU_FILL);
			gluQuadricNormals (shapes, GLU_SMOOTH);
			//gluSphere(shapes, power(grow,i/ribs), 5, 5);
			gluSphere (shapes, exp (i / ribs * log (grow)), 6, 4);

			break;

		case LIGHTMAP:
		case POINTS:
			glBegin (GL_POINTS);
			glVertex3d (0, 0, 0);
			glEnd ();

			break;
		}

		glPopMatrix ();
	}
}

void hack_reshape (xstuff_t * XStuff)
{
	glViewport (0, 0, (GLsizei) XStuff->windowWidth, (GLsizei) XStuff->windowHeight);
	glMatrixMode (GL_PROJECTION);
	glLoadIdentity ();
	gluPerspective (30, (float)XStuff->windowWidth / (float)XStuff->windowHeight, 100, 300);
}

void hack_init (xstuff_t * XStuff)	// Called right after the window is created, and OpenGL is initialized.
{
	float glfLightPosition[4] = { 100.0, 100.0, 100.0, 0.0 };
	float glfFog[4] = { 0.0, 0.0, 0.3, 1.0 };
	float DiffuseLightColor[4] = { 1, 0.8, 0.4, 1.0 };
	int i, j;
	double x, y, r, d;
	unsigned char texbuf[64][64];

	glEnable (GL_DEPTH_TEST);

	glBlendFunc (GL_SRC_ALPHA, GL_ONE_MINUS_SRC_ALPHA);
	glEnable (GL_BLEND);

	if ((dGeometry == SPHERES) || (dGeometry == BIGSPHERES)) {
		glEnable (GL_CULL_FACE);
		glCullFace (GL_BACK);

		glLightfv (GL_LIGHT0, GL_POSITION, glfLightPosition);
		glEnable (GL_LIGHTING);
		glEnable (GL_LIGHT0);

		glLightfv (GL_LIGHT0, GL_DIFFUSE, DiffuseLightColor);

		shapes = gluNewQuadric ();
	} else if (dGeometry == POINTS) {
		glHint (GL_POINT_SMOOTH_HINT, GL_NICEST);
		glPointSize (3);
		glEnable (GL_POINT_SMOOTH);

		glFogi (GL_FOG_MODE, GL_LINEAR);
		glHint (GL_FOG_HINT, GL_NICEST);
		glFogfv (GL_FOG_COLOR, glfFog);
		glFogf (GL_FOG_START, 150);
		glFogf (GL_FOG_END, 250);
		glEnable (GL_FOG);
	} else if (dGeometry == LIGHTMAP) {
		r = 32;

		memset ((void *)&texbuf, 0, 4096);

		for (i = 0; i < 64; i++) {
			for (j = 0; j < 64; j++) {
				x = abs (i - 32);
				y = abs (j - 32);
				d = sqrt (x * x + y * y);

				if (d < r) {
					d = 1 - (d / r);
					texbuf[i][j] = (char)(255.0f * d * d);
				}
			}
		}

		gluBuild2DMipmaps (GL_TEXTURE_2D, 1, 64, 64, GL_LUMINANCE, GL_UNSIGNED_BYTE, texbuf);

		glTexParameteri (GL_TEXTURE_2D, GL_TEXTURE_MIN_FILTER, GL_LINEAR_MIPMAP_NEAREST);
		glTexParameteri (GL_TEXTURE_2D, GL_TEXTURE_MAG_FILTER, GL_LINEAR);
	}

	hack_reshape (XStuff);
}

void hack_cleanup (xstuff_t * XStuff)
{
	if ((dGeometry == SPHERES) || (dGeometry == BIGSPHERES))
		gluDeleteQuadric(shapes);
}

void hack_draw (xstuff_t * XStuff, double currentTime, float lastFrameDuration)
{
	if (dGeometry == LIGHTMAP) {
		int i, j, offset;
		float fb_buffer[NRPOINTS * NRLINES * 3];
		double x, y, w;

		glDisable (GL_DEPTH_TEST);
		glFeedbackBuffer (NRPOINTS * NRLINES * 3, GL_2D, fb_buffer);
		glRenderMode (GL_FEEDBACK);

		glMatrixMode (GL_MODELVIEW);
		glPushMatrix ();
		glLoadIdentity ();

		gluLookAt (200, 0, 0, 0, 0, 0, 0, 0, 1);

		glRotated (90, 0, 1, 0);
		glRotated (10 * currentTime, 0, 0, 1);

		glNewList (1, GL_COMPILE);
		Horn (NRPOINTS, 315 * sin (currentTime * 0.237), 45 /* + 20 * sin(currentTime * 0.133) */ , 1500 * sin (currentTime * 0.213), 7.5 + 2.5 * sin (currentTime * 0.173), 1);
		glEndList ();

		if (dOffAngle)
			glRotated (45, 1, 1, 0);

		for (i = 0; i < NRLINES; i++) {
			glPushMatrix ();

			glRotated (360 * i / NRLINES, 0, 0, 1);
			glRotated (45, 1, 0, 0);
			glTranslated (0, 0, 2);

			glCallList (1);

			glPopMatrix ();
		}

		glPopMatrix ();

		glRenderMode (GL_RENDER);

		glTexEnvi (GL_TEXTURE_ENV, GL_TEXTURE_ENV_MODE, GL_MODULATE);

		glMatrixMode (GL_PROJECTION);
		glPushMatrix ();
		glLoadIdentity ();

		glMatrixMode (GL_MODELVIEW);
		glPushMatrix ();
		glLoadIdentity ();

		glScalef (2.0 / XStuff->windowWidth, 2.0 / XStuff->windowHeight, 1.0);
		glTranslatef (-0.5 * XStuff->windowWidth, -0.5 * XStuff->windowHeight, 0.0);

		glClear (GL_COLOR_BUFFER_BIT);

		glEnable (GL_TEXTURE_2D);
		glBlendFunc (GL_ONE, GL_ONE);
		glEnable (GL_BLEND);

		glBegin (GL_QUADS);

		w = XStuff->windowWidth >> 7;

		for (j = 0; j < NRPOINTS; j++) {
			glColor3f (0.25 - 0.15 * j / NRPOINTS, 0.15 + 0.10 * j / NRPOINTS, 0.06 + 0.14 * j / NRPOINTS);
			for (i = 0; i < NRLINES; i++) {
				offset = 3 * (i * NRPOINTS + j);
				x = fb_buffer[offset + 1];
				y = fb_buffer[offset + 2];

				glTexCoord2f (0.0, 0.0);
				glVertex2f (x - w, y - w);

				glTexCoord2f (1.0, 0.0);
				glVertex2f (x + w, y - w);

				glTexCoord2f (1.0, 1.0);
				glVertex2f (x + w, y + w);

				glTexCoord2f (0.0, 1.0);
				glVertex2f (x - w, y + w);
			}
		}

		glEnd ();

		glMatrixMode (GL_MODELVIEW);
		glPopMatrix ();
		glMatrixMode (GL_PROJECTION);
		glPopMatrix ();
	} else {
		int i;

		glClear (GL_DEPTH_BUFFER_BIT | GL_COLOR_BUFFER_BIT);

		glMatrixMode (GL_MODELVIEW);
		glLoadIdentity ();
		gluLookAt (200, 0, 0, 0, 0, 0, 0, 0, 1);

		glPushMatrix ();
		glRotated (90, 0, 1, 0);
		glRotated (10 * currentTime, 0, 0, 1);

		glColor3d (0.9, 0.7, 0.5);

		glNewList (1, GL_COMPILE);
		Horn (NRPOINTS, 315 * sin (currentTime * 0.237), 50 + 20 * sin (currentTime * 0.133), 1500 * sin (currentTime * 0.213), 7.5 + 2.5 * sin (currentTime * 0.173), 0.6 + 0.1 * sin (currentTime * 0.317));
		glEndList ();

		if (dOffAngle)
			glRotated (45, 1, 1, 0);

		for (i = 0; i < NRLINES; i++) {
			glPushMatrix ();
			glRotated (360 * i / NRLINES, 0, 0, 1);
			glRotated (45, 1, 0, 0);
			glTranslated (0, 0, 2);
			glCallList (1);
			glPopMatrix ();
		}

		glPopMatrix ();
	}
}

void setDefaults (int preset)
{
	dGeometry = preset;

	switch (preset) {
	case TRIANGLES:
		NRPOINTS = 80;
		NRLINES = 32;

		break;
	case SPHERES:
		NRPOINTS = 80;
		NRLINES = 30;

		break;
	case BIGSPHERES:
		NRPOINTS = 20;
		NRLINES = 5;

		break;
	case POINTS:
		NRPOINTS = 250;
		NRLINES = 30;

		break;
	case LIGHTMAP:
		NRPOINTS = 150;
		NRLINES = 20;

		break;
	}
}

void hack_handle_opts (int argc, char **argv)
{
	int change_flag = 0;

	setDefaults(LIGHTMAP);
	dOffAngle = 0;

	while (1) {
		int c;

#ifdef HAVE_GETOPT_H
		static struct option long_options[] = {
			{"help", 0, 0, 'h'},
			DRIVER_OPTIONS_LONG {"preset", 1, 0, 'P'},
			{"triangles", 0, 0, 0},
			{"spheres", 0, 0, 1},
			{"bigspheres", 0, 0, 2},
			{"points", 0, 0, 3},
			{"lightmap", 0, 0, 4},
			{"lines", 1, 0, 'l'},
			{"points", 1, 0, 'p'},
			{"offangle", 0, 0, 'o'},
			{0, 0, 0, 0}
		};

		c = getopt_long (argc, argv, DRIVER_OPTIONS_SHORT "hP:l:p:o", long_options, NULL);
#else
		c = getopt (argc, argv, DRIVER_OPTIONS_SHORT "hP:l:p:o");
#endif
		if (c == -1)
			break;

		switch (c) {
			DRIVER_OPTIONS_CASES case 'h':printf ("%s:"
#ifndef HAVE_GETOPT_H
							      " Not built with GNU getopt.h, long options *NOT* enabled."
#endif
							      "\n" DRIVER_OPTIONS_HELP "\t--preset/-P <arg>\n" "\t--triangles\n"
							      "\t--spheres\n" "\t--bigspheres\n" "\t--points\n" "\t--lightmap\n" 
							      "\t--lines/-l <arg>\n" "\t--points/-p <arg>\n" "\t--offangle\n", argv[0]);
			exit (1);
		case 'P':
			change_flag = 1;
			setDefaults (strtol_minmaxdef (optarg, 10, 1, 5, 0, 1, "--preset: "));
			break;
		case 0:
		case 1:
		case 2:
		case 3:
		case 4:
			change_flag = 1;
			setDefaults (c + 1);
			break;
		case 'l':
			change_flag = 1;
			NRLINES = strtol_minmaxdef (optarg, 10, 1, 1024, 0, 1, "--lines: ");
			break;
		case 'p':
			change_flag = 1;
			NRPOINTS = strtol_minmaxdef (optarg, 10, 1, 4096, 0, 1, "--points: ");
			break;
		case 'o':
			change_flag = 1;
			dOffAngle = 1;
		}
	}

	if (!change_flag) {
		dGeometry = rsRandi (5) + 1;
		dOffAngle = rsRandi (2);

		switch (dGeometry) {
		case TRIANGLES:
		case SPHERES:
			NRLINES = rsRandi (7) + 2;
			NRPOINTS = rsRandi (96) + 32;
			break;

		case BIGSPHERES:
			NRLINES = rsRandi (7) + 4;
			NRPOINTS = rsRandi (32) + 4;
			break;

		case POINTS:
		case LIGHTMAP:
			NRLINES = rsRandi (32) + 4;
			NRPOINTS = rsRandi (64) + 64;
		}
	}
}

